pal/input/web/gamepad-input.ts

```diff
@@ -233,11 +294,130 @@ export class GamepadInputDevice {
         }
         // update cache
         GamepadInputDevice._cachedWebGamepads = webGamepads;
+    }
+
+    private static _scanGamepads () {
+        const devices: GamepadInputDevice[] = [];
+        GamepadInputDevice._scanWebGamepads(devices);
+        GamepadInputDevice._scanWebXRGamepads(devices);
         // emit event
         for (let i = 0; i < devices.length; ++i) {
             const device = devices[i];
             GamepadInputDevice._eventTarget.emit(InputEventType.GAMEPAD_INPUT, new EventGamepad(InputEventType.GAMEPAD_INPUT, device));
         }
+        GamepadInputDevice._scanWebXRGamepadsPose();
+    }
+
+    private static _scanWebXRGamepads (devices: GamepadInputDevice[]) {
+        const webxrGamepadMap = GamepadInputDevice._getWebXRGamepadMap();
+        if (!webxrGamepadMap) {
+            // update cache
+            GamepadInputDevice._cachedWebXRGamepadMap = null;
+            if (GamepadInputDevice.xr && GamepadInputDevice.xr._connected) {
+                GamepadInputDevice.xr._connected = false;
+                GamepadInputDevice._eventTarget.emit(InputEventType.GAMEPAD_CHANGE,
+                    new EventGamepad(InputEventType.GAMEPAD_CHANGE, GamepadInputDevice.xr));
+                devices.push(GamepadInputDevice.xr);
+            }
+            return;
+        }
+
+        if (!GamepadInputDevice.xr) {
+            // webxr gamepads index is -1 https://www.w3.org/TR/webxr-gamepads-module-1/#navigator-differences
+            GamepadInputDevice.xr = new GamepadInputDevice(-1);
+        }
+
+        const left = webxrGamepadMap.get(XRLeftHandedness);
+        const right = webxrGamepadMap.get(XRRightHandedness);
+        if (!left && !right) {
+            if (GamepadInputDevice.xr._connected) {
+                GamepadInputDevice.xr._connected = false;
+                GamepadInputDevice._eventTarget.emit(InputEventType.GAMEPAD_CHANGE,
+                    new EventGamepad(InputEventType.GAMEPAD_CHANGE, GamepadInputDevice.xr));
+            }
+        } else if (!GamepadInputDevice.xr._connected) {
+            GamepadInputDevice.xr._connected = true;
+            GamepadInputDevice._eventTarget.emit(InputEventType.GAMEPAD_CHANGE,
+                new EventGamepad(InputEventType.GAMEPAD_CHANGE, GamepadInputDevice.xr));
+        }
+
+        if (GamepadInputDevice.checkGamepadChanged(left,
+            GamepadInputDevice._cachedWebXRGamepadMap?.get(XRLeftHandedness))) {
+            devices.push(GamepadInputDevice.xr);
+        } else if (GamepadInputDevice.checkGamepadChanged(right,
+            GamepadInputDevice._cachedWebXRGamepadMap?.get(XRRightHandedness))) {
+            devices.push(GamepadInputDevice.xr);
```

Comment:
Push the same value `GamepadInputDevice.xr` twice?

Reply:
Here, as long as there is a change in data on the left or right gamepad, push once instead of pushing twice.